### ISSUE TITLE
Struct#dig uses exceptions for control-flow

### ISSUE THREAD

**Issue by chrisseaton** [oracle/truffleruby#2299]
https://github.com/oracle/truffleruby/blob/d3ff03e9fd13212a69657a2b0c3ec5ff1727559d/src/main/ruby/truffleruby/core/struct.rb#L257-L261

Must be a better way?

Discovered with @jantnovi.

**Comment by eregon** [comment]
Right, we can probably inline Struct#[] and adapt it as needed.

**Comment by chrisseaton** [comment]
Started investigating this. What is going on? How on earth is TruffleRuby so good at removing exceptions here?!

```ruby
require 'benchmark/ips'

puts RUBY_DESCRIPTION

struct = Struct.new(:foo).new(14)

Benchmark.ips do |x|
  x.report('found') do
      struct[:foo]
  end

  x.report('not-found') do
    struct[:bar]
  rescue IndexError, NameError
    nil
  end
end
```

```
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]
Warming up --------------------------------------
               found     1.314M i/100ms
           not-found    73.002k i/100ms
Calculating -------------------------------------
               found     13.018M (± 3.2%) i/s -     65.682M in   5.050757s
           not-found    829.543k (± 3.0%) i/s -      4.161M in   5.020941s

truffleruby 21.1.0-dev-ea1c4dc3, like ruby 2.7.2, Interpreted JVM [x86_64-darwin]
Warming up --------------------------------------
               found    62.439k i/100ms
           not-found     4.818k i/100ms
Calculating -------------------------------------
               found    729.528k (± 5.5%) i/s -      3.684M in   5.066823s
           not-found     73.650k (± 6.6%) i/s -    366.168k in   5.003143s

truffleruby 21.1.0-dev-ea1c4dc3, like ruby 2.7.2, GraalVM CE JVM [x86_64-darwin]
Warming up --------------------------------------
               found    56.361M i/100ms
           not-found   129.870k i/100ms
Calculating -------------------------------------
               found    107.015M (±15.9%) i/s -    563.605M in   5.460417s
           not-found     99.236M (± 5.8%) i/s -    493.636M in   4.993145s
```

The exception does seem to exist:

```ruby
struct = Struct.new(:foo).new(14)

begin
  p struct[:bar]
rescue IndexError, NameError
  p 'rescued'
  p $!
end

p $!
```

```
chrisseaton@Chris-Seatons-MacBook-Pro truffleruby % ruby test.rb                
"rescued"
#<NameError: no member 'bar' in struct>
nil
chrisseaton@Chris-Seatons-MacBook-Pro truffleruby % jt --use jvm-ce ruby test.rb
Using TruffleRuby with Graal: mxbuild/truffleruby-jvm-ce
$ /Users/chrisseaton/Documents/truffleruby/mxbuild/truffleruby-jvm-ce/languages/ruby/bin/ruby \
  --experimental-options \
  --core-load-path=/Users/chrisseaton/Documents/truffleruby/src/main/ruby/truffleruby \
  test.rb
"rescued"
#<NameError: no member 'bar' in struct>
nil
```

Am I out of date with the performance of exceptions? TruffleRuby can remove creating the backtrace? And what isn't `$!` set in either MRI or TruffleRuby at the end of the program?

**Comment by eregon** [comment]
> TruffleRuby can remove creating the backtrace?

Yes, as long as we can see `$!` isn't used in the `rescue` body.
A call in rescue would invalidate that, as the method called might access `$!`.

`$!` is only set in `rescue` and is cleared after. Since the `rescue` rescued the exception, there is no more an ongoing unwinding exception.

**Comment by eregon** [comment]
Also AbstractTruffleException builds its stacktrace lazily as unwinds.
We need to compute the stacktrace eagerly only if the Ruby Exception is captured (e.g. in $!) and no longer unwinding the stack.

**Comment by chrisseaton** [comment]
Wow I'm impressed we got that far.

So there's no issue here is there? Shall we close it?

**Comment by eregon** [comment]
It relies on inlining Struct#dig -> Struct#[] -> #raise to optimize the exception away entirely but that seems reasonable and likely.

In interpreter and first tier it would likely still be faster to avoid the exception.
I'm not sure how frequent it is to `dig` on a Struct though.

**Comment by chrisseaton** [comment]
Deferred until after #2301.
